_posts/2024-02-09-Database-interaction-with-skunk-scala-library:-A-comprehensive-guide.md

```diff
@@ -0,0 +1,555 @@
+                } yield ()
+              } *>
+              resources.use { resource =>
+                resource.use { session =>
+                  for {
+                    _           <- IO(println("Using a pooled session..."))
+                    dateAndTime <- session.unique(sql"select current_timestamp".query(timestamptz))
+                    _           <- IO(println(s"Current date and time is $dateAndTime."))
+                  } yield ExitCode.Success
+                }
+              }
+      }
+  ```
+  We first load configuration values from `application.conf`. If values are properly set up, we create both single and pooled database connections with the help of our `DbConnection` class we previously created. Otherwise, we print errors to the console to better understand what went wrong. We now test our database connections by fetching the current date and time of our system, and if our connections was succesfully established, we should have an output similar to the following.
+  ```
+  Using a single session...
+  Current date and time is 2024-02-20T17:29:20.972023Z.
+  Using a pooled session...
+  Current date and time is 2024-02-20T17:29:21.058790Z.
+  ```
+  `Session.single` and `Session.pooled` are two distinct methods for establishing database connections, each with its own approach for managing connections. The following comparison can assist you in selecting the method that is most suitable for your application's needs and scalability requirements. For simplicity, we will use `Session.single` throughout, but feel free to experiment with `Session.pooled` to understand its benefits.
+
+  |Feature        | Session.single | Session.pooled |
+  |---------------|----------------|----------------|
+  |Connection Management | Creates a single connection per session. | Manages a pool of connections.
+  |Suitable For | Low database load or manual connection management. | High database load or automatic connection pooling.
+  |Connection Reuse  |Each call establishes a new connection. | Reuses existing connections from the pool.
+  |Performance | May have higher overhead due to frequent connection           establishment. | Generally more efficient due to connection pooling.
+
+## 7. Type Mapping in Skunk
+
+Let's talk about type mapping as it is an essential concept that helps us in the execution of skunk queries and commands. It's a process of converting between PostgreSQL data types and corresponding Scala types. This mapping ensures that data retrieved from the database can be represented and manipulated in a type-safe manner within Scala code. Through the use of `Codec`(encoder and decoder), we can create a mapping between fields from our `users` table and values from our `User` case class as follow.
+```scala
+    object main extends IOApp {
+      val codec: Codec[User] =
+        (uuid ~ varchar ~ varchar).imap { 
+          case id ~ name ~ email => User(id, name, email)
+        }(user => user.id ~ user.name ~ user.email)
```

Comment:
the `~` operator seems to be deprecated, let's create simple tuples

Reply:
Would you like me to provide an example for those using Scala 2 and Skunk <0.6?